Program Deep Link Throws If Not Already Logged In

**Describe the bug**
This started after the switch to enable BYPASS_LOGIN_LANGUAGE_SCREENS. When a user loads civiform from a program deep link an exception is occurring if they are not already logged in as a guest or with an account. 

Prior to the new login format the redirect controller would handle this by going to the login page. I think [that this code](https://github.com/civiform/civiform/blob/main/server/app/controllers/applicant/RedirectController.java#L80-L83) would still work if there is a new controller action to redirect to, but I don't know which that is.

**To Reproduce**
Steps to reproduce the behavior:

1. Open a new private browser window that has not been to https://civiform.seattle.gov
2. In the browser address bar paste https://civiform.seattle.gov/programs/upcoming and hit enter to load the page
3. Receive an exception page (See below for details)

For local development just create and publish a program first and use that deep link. 

**Workaround**

1. Go to https://civiform.seattle.gov, this will trigger a guest account
2. In the browser address bar paste https://civiform.seattle.gov/programs/upcoming and hit enter to load the page
3. Page will load

**Expected behavior**
The deep link should load whether or not the client was already logged in as guest or with a user account.

**Exception**
```
@81hn0ahc9 - Internal server error, for (GET) [/programs/upcoming]

play.api.UnexpectedException: Unexpected exception[NoSuchElementException: No value present]
 at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:358)
 at play.api.http.HttpErrorHandlerExceptions.throwableToUsefulException(HttpErrorHandler.scala)
 at play.http.DefaultHttpErrorHandler.throwableToUsefulException(DefaultHttpErrorHandler.java:246)
 at play.http.DefaultHttpErrorHandler.onServerError(DefaultHttpErrorHandler.java:182)
 at controllers.ErrorHandler.onServerError(ErrorHandler.java:102)
 at play.core.j.JavaHttpErrorHandlerAdapter.onServerError(JavaHttpErrorHandlerAdapter.scala:32)
 at play.core.server.AkkaHttpServer$$anonfun$2.applyOrElse(AkkaHttpServer.scala:441)
 at play.core.server.AkkaHttpServer$$anonfun$2.applyOrElse(AkkaHttpServer.scala:433)
 at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:490)
 at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:63)
 at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:100)
 at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
 at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94)
 at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:100)
 at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:49)
 at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48)
 at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
 at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
 at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
 at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)

Caused by: java.util.NoSuchElementException: No value present
 at java.base/java.util.Optional.get(Unknown Source)
 at controllers.applicant.RedirectController.programBySlug(RedirectController.java:89)
 at router.Routes$$anonfun$routes$1.$anonfun$applyOrElse$303(Routes.scala:3622)
 at play.core.routing.HandlerInvokerFactory$$anon$9.resultCall(HandlerInvoker.scala:154)
 at play.core.routing.HandlerInvokerFactory$JavaActionInvokerFactory$$anon$3$$anon$4$$anon$5.invocation(HandlerInvoker.scala:115)
 at play.core.j.JavaAction$$anon$1.call(JavaAction.scala:119)
 at play.http.DefaultActionCreator$1.call(DefaultActionCreator.java:33)
 at play.core.j.JavaAction.$anonfun$apply$8(JavaAction.scala:175)
 at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:678)
 at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:467)
 at play.core.j.HttpExecutionContext.$anonfun$execute$1(HttpExecutionContext.scala:64)
 at play.api.libs.streams.Execution$trampoline$.executeScheduled(Execution.scala:109)
 at play.api.libs.streams.Execution$trampoline$.execute(Execution.scala:71)
 at scala.concurrent.impl.Promise$Transformation.submitWithValue(Promise.scala:429)
 at scala.concurrent.impl.Promise$DefaultPromise.submitWithValue(Promise.scala:338)
 at scala.concurrent.impl.Promise$DefaultPromise.dispatchOrAddCallbacks(Promise.scala:312)
 at scala.concurrent.impl.Promise$DefaultPromise.flatMap(Promise.scala:176)
 at play.api.libs.streams.Accumulator$.$anonfun$futureToSink$1(Accumulator.scala:231)
 at akka.stream.impl.Transform.apply(TraversalBuilder.scala:159)
 at akka.stream.impl.PhasedFusingActorMaterializer.materialize(PhasedFusingActorMaterializer.scala:522)
 at play.core.server.AkkaHttpServer$$anonfun$2.applyOrElse(AkkaHttpServer.scala:441)
 at play.core.server.AkkaHttpServer$$anonfun$2.applyOrElse(AkkaHttpServer.scala:433)
 at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:490)
 at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:63)
 at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:100)
 at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
 at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94)
 at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:100)
 at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:49)
 at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48)
 at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
 at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
 at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
 at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)

Caused by: java.util.NoSuchElementException: No value present
 at java.base/java.util.Optional.get(Unknown Source)
 at controllers.applicant.RedirectController.programBySlug(RedirectController.java:89)
 at router.Routes$$anonfun$routes$1.$anonfun$applyOrElse$303(Routes.scala:3622)
 at play.core.routing.HandlerInvokerFactory$$anon$9.resultCall(HandlerInvoker.scala:154)
 at play.core.routing.HandlerInvokerFactory$JavaActionInvokerFactory$$anon$3$$anon$4$$anon$5.invocation(HandlerInvoker.scala:115)
 at play.core.j.JavaAction$$anon$1.call(JavaAction.scala:119)
 at play.http.DefaultActionCreator$1.call(DefaultActionCreator.java:33)
 at play.core.j.JavaAction.$anonfun$apply$8(JavaAction.scala:175)
 at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:678)
 at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:467)
 at play.core.j.HttpExecutionContext.$anonfun$execute$1(HttpExecutionContext.scala:64)
 at play.api.libs.streams.Execution$trampoline$.executeScheduled(Execution.scala:109)
 at play.api.libs.streams.Execution$trampoline$.execute(Execution.scala:71)
 at scala.concurrent.impl.Promise$Transformation.submitWithValue(Promise.scala:429)
 at scala.concurrent.impl.Promise$DefaultPromise.submitWithValue(Promise.scala:338)
 at scala.concurrent.impl.Promise$DefaultPromise.dispatchOrAddCallbacks(Promise.scala:312)
 at scala.concurrent.impl.Promise$DefaultPromise.flatMap(Promise.scala:176)
 at play.api.libs.streams.Accumulator$.$anonfun$futureToSink$1(Accumulator.scala:231)
 at akka.stream.impl.Transform.apply(TraversalBuilder.scala:159)
 at akka.stream.impl.PhasedFusingActorMaterializer.materialize(PhasedFusingActorMaterializer.scala:522)
```

--- a/browser-test/src/program_deep_link.test.ts
+++ b/browser-test/src/program_deep_link.test.ts
@@ -5,13 +5,15 @@ import {
   loginAsAdmin,
   loginAsGuest,
   loginAsTestUser,
-  logout, resetContext,
+  logout,
+  resetContext,
   selectApplicantLanguage,
   validateScreenshot,
+  TestContext,
 } from './support'
 
 describe('navigating to a deep link', () => {
-  const ctx = createTestContext()
+  const ctx: TestContext = createTestContext()
 
   const questionText = 'What is your address?'
 
@@ -70,7 +72,7 @@ describe('navigating to a deep link', () => {
 
   it('takes guests and logged in users through the flow correctly', async () => {
     await resetContext(ctx)
-    ctx.browser.close()
+    await ctx.browser.close()
     const {page} = ctx
 
     // Exercise guest path

--- a/browser-test/src/program_deep_link.test.ts
+++ b/browser-test/src/program_deep_link.test.ts
@@ -5,7 +5,7 @@ import {
   loginAsAdmin,
   loginAsGuest,
   loginAsTestUser,
-  logout,
+  logout, resetContext,
   selectApplicantLanguage,
   validateScreenshot,
 } from './support'
@@ -43,6 +43,7 @@ describe('navigating to a deep link', () => {
   })
 
   it('shows a login prompt for guest users', async () => {
+    await resetContext(ctx)
     const {page} = ctx
 
     await gotoEndpoint(page, '/programs/test-deep-link')
@@ -57,6 +58,7 @@ describe('navigating to a deep link', () => {
   })
 
   it('does not show login prompt for logged in users', async () => {
+    await resetContext(ctx)
     const {page} = ctx
 
     await gotoEndpoint(page, '/programs/test-deep-link')
@@ -67,6 +69,8 @@ describe('navigating to a deep link', () => {
   })
 
   it('takes guests and logged in users through the flow correctly', async () => {
+    await resetContext(ctx)
+    ctx.browser.close()
     const {page} = ctx
 
     // Exercise guest path

--- a/browser-test/src/program_deep_link.test.ts
+++ b/browser-test/src/program_deep_link.test.ts
@@ -72,7 +72,6 @@ describe('navigating to a deep link', () => {
 
   it('takes guests and logged in users through the flow correctly', async () => {
     await resetContext(ctx)
-    await ctx.browser.close()
     const {page} = ctx
 
     // Exercise guest path
@@ -101,5 +100,26 @@ describe('navigating to a deep link', () => {
     expect(await page.innerText('.cf-applicant-question-text')).toContain(
       questionText,
     )
+  })
+
+  it('Non-logged in user should get redirected to the program page and not an error', async () => {
+    await resetContext(ctx)
+    const {page, browserContext} = ctx
+
+    await selectApplicantLanguage(page, 'English')
+
+    await logout(page)
+    await browserContext.clearCookies()
+    await gotoEndpoint(page, '/programs/test-deep-link')
+    await page.click('text="Continue to application"')
+    await selectApplicantLanguage(page, 'English')
+
+    // Assert
+    await page.click('#continue-application-button')
+    expect(await page.innerText('.cf-applicant-question-text')).toContain(
+      questionText,
+    )
+
+    await logout(page)
   })
 })

--- a/browser-test/src/support/index.ts
+++ b/browser-test/src/support/index.ts
@@ -185,9 +185,6 @@ export interface TestContext {
  *     so none of its properties should be cached and reused between tests.
  */
 export const createTestContext = (clearDb = true): TestContext => {
-  // let browser: Browser
-  // let browserContext: BrowserContext
-
   // TestContext properties are set in resetContext() later. For now we just
   // need an object that we can return to caller. Caller is expected to access
   // it only from before/afterX functions or tests.

--- a/browser-test/src/support/index.ts
+++ b/browser-test/src/support/index.ts
@@ -193,7 +193,6 @@ export const createTestContext = (clearDb = true): TestContext => {
   // it only from before/afterX functions or tests.
   const ctx: TestContext = {} as unknown as TestContext
 
-
   beforeAll(async () => {
     ctx.browser = await chromium.launch()
     await resetContext(ctx)

--- a/browser-test/src/support/index.ts
+++ b/browser-test/src/support/index.ts
@@ -140,6 +140,8 @@ export interface TestContext {
    * Methods: https://playwright.dev/docs/api/class-page
    */
   page: Page
+  browser: Browser
+  browserContext: BrowserContext
   browserErrorWatcher: BrowserErrorWatcher
   useMobile: boolean
 
@@ -183,55 +185,18 @@ export interface TestContext {
  *     so none of its properties should be cached and reused between tests.
  */
 export const createTestContext = (clearDb = true): TestContext => {
-  let browser: Browser
-  let browserContext: BrowserContext
+  // let browser: Browser
+  // let browserContext: BrowserContext
 
   // TestContext properties are set in resetContext() later. For now we just
   // need an object that we can return to caller. Caller is expected to access
   // it only from before/afterX functions or tests.
   const ctx: TestContext = {} as unknown as TestContext
 
-  // We create new browser context and session before each test. It's
-  // important to get fresh browser context so that each test gets its own
-  // video file. If we reuse same browser context across multiple test cases -
-  // we'll get one huge video for all tests.
-  async function resetContext() {
-    if (browserContext != null) {
-      try {
-        if (!DISABLE_BROWSER_ERROR_WATCHER) {
-          ctx.browserErrorWatcher.failIfContainsErrors()
-        }
-      } finally {
-        // browserErrorWatcher might throw an error that should bubble up all
-        // the way to the developer. Regardless whether the error is thrown or
-        // not we need to close the browser context. Without that some processes
-        // won't be finished, like saving videos.
-        await browserContext.close()
-      }
-    }
-    browserContext = await makeBrowserContext(browser, ctx.useMobile)
-    ctx.page = await browserContext.newPage()
-    ctx.browserErrorWatcher = new BrowserErrorWatcher(ctx.page)
-    // Default timeout is 30s. It's too long given that civiform is not JS
-    // heavy and all elements render quite quickly. Setting it to 5 sec so that
-    // tests fail fast.
-    ctx.page.setDefaultTimeout(8000)
-    ctx.adminQuestions = new AdminQuestions(ctx.page)
-    ctx.adminPrograms = new AdminPrograms(ctx.page)
-    ctx.adminApiKeys = new AdminApiKeys(ctx.page)
-    ctx.adminProgramStatuses = new AdminProgramStatuses(ctx.page)
-    ctx.applicantQuestions = new ApplicantQuestions(ctx.page)
-    ctx.adminPredicates = new AdminPredicates(ctx.page)
-    ctx.adminTranslations = new AdminTranslations(ctx.page)
-    ctx.tiDashboard = new TIDashboard(ctx.page)
-    ctx.adminTiGroups = new AdminTIGroups(ctx.page)
-    await ctx.page.goto(BASE_URL)
-    await closeWarningMessage(ctx.page)
-  }
 
   beforeAll(async () => {
-    browser = await chromium.launch()
-    await resetContext()
+    ctx.browser = await chromium.launch()
+    await resetContext(ctx)
     // clear DB at beginning of each test suite. While data can leak/share
     // between test cases within a test file, data should not be shared
     // between test files.
@@ -240,7 +205,7 @@ export const createTestContext = (clearDb = true): TestContext => {
   })
 
   beforeEach(async () => {
-    await resetContext()
+    await resetContext(ctx)
   })
 
   afterEach(async () => {
@@ -250,14 +215,52 @@ export const createTestContext = (clearDb = true): TestContext => {
     // resetting context here so that afterAll() functions of current describe()
     // block and beforeAll() functions of the next describe() block have fresh
     // result.page object.
-    await resetContext()
+    await resetContext(ctx)
   })
 
   afterAll(async () => {
-    await endSession(browser)
+    await endSession(ctx.browser)
   })
 
   return ctx
+}
+
+// We create new browser context and session before each test. It's
+// important to get fresh browser context so that each test gets its own
+// video file. If we reuse same browser context across multiple test cases -
+// we'll get one huge video for all tests.
+export async function resetContext(ctx: TestContext) {
+  if (ctx.browserContext != null) {
+    try {
+      if (!DISABLE_BROWSER_ERROR_WATCHER) {
+        ctx.browserErrorWatcher.failIfContainsErrors()
+      }
+    } finally {
+      // browserErrorWatcher might throw an error that should bubble up all
+      // the way to the developer. Regardless whether the error is thrown or
+      // not we need to close the browser context. Without that some processes
+      // won't be finished, like saving videos.
+      await ctx.browserContext.close()
+    }
+  }
+  ctx.browserContext = await makeBrowserContext(ctx.browser, ctx.useMobile)
+  ctx.page = await ctx.browserContext.newPage()
+  ctx.browserErrorWatcher = new BrowserErrorWatcher(ctx.page)
+  // Default timeout is 30s. It's too long given that civiform is not JS
+  // heavy and all elements render quite quickly. Setting it to 5 sec so that
+  // tests fail fast.
+  ctx.page.setDefaultTimeout(8000)
+  ctx.adminQuestions = new AdminQuestions(ctx.page)
+  ctx.adminPrograms = new AdminPrograms(ctx.page)
+  ctx.adminApiKeys = new AdminApiKeys(ctx.page)
+  ctx.adminProgramStatuses = new AdminProgramStatuses(ctx.page)
+  ctx.applicantQuestions = new ApplicantQuestions(ctx.page)
+  ctx.adminPredicates = new AdminPredicates(ctx.page)
+  ctx.adminTranslations = new AdminTranslations(ctx.page)
+  ctx.tiDashboard = new TIDashboard(ctx.page)
+  ctx.adminTiGroups = new AdminTIGroups(ctx.page)
+  await ctx.page.goto(BASE_URL)
+  await closeWarningMessage(ctx.page)
 }
 
 export const endSession = async (browser: Browser) => {

--- a/server/app/controllers/applicant/RedirectController.java
+++ b/server/app/controllers/applicant/RedirectController.java
@@ -5,6 +5,7 @@ import static controllers.CallbackController.REDIRECT_TO_SESSION_KEY;
 import static views.components.ToastMessage.ToastType.ALERT;
 
 import auth.CiviFormProfile;
+import auth.GuestClient;
 import auth.ProfileUtils;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
@@ -77,11 +78,14 @@ public final class RedirectController extends CiviFormController {
   public CompletionStage<Result> programBySlug(Http.Request request, String programSlug) {
     Optional<CiviFormProfile> profile = profileUtils.currentUserProfile(request);
 
-    if (!featureFlags.getFlagEnabled(request, FeatureFlag.BYPASS_LOGIN_LANGUAGE_SCREENS)
-        && profile.isEmpty()) {
-      Result result = redirect(routes.HomeController.loginForm(Optional.of("login")));
+    if (profile.isEmpty()) {
+      Result result;
+      if (featureFlags.getFlagEnabled(request, FeatureFlag.BYPASS_LOGIN_LANGUAGE_SCREENS)) {
+        result = redirect(routes.CallbackController.callback(GuestClient.CLIENT_NAME).url());
+      } else {
+        result = redirect(routes.HomeController.loginForm(Optional.of("login")));
+      }
       result = result.withSession(ImmutableMap.of(REDIRECT_TO_SESSION_KEY, request.uri()));
-
       return CompletableFuture.completedFuture(result);
     }
 

